### PR TITLE
PR candidate: Adding #define preprocessor directive

### PR DESCRIPTION
**Pull request by gg64du02**
Define blocks are not yet included.

### CODE REVIEW

File: Source/MDK-Complete.sln [hunk 1]
```diff
@@ -1,84 +1,92 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Mdk.Extractor", "Mdk.Extractor\Mdk.Extractor.csproj", "{88AD81CF-B0C0-4D8A-B682-FC0133258810}"
```
Comment (malware-dev) [inline Source/MDK-Complete.sln:3]: What's with all the changes in _this_ file?

Comment (gg64du02) [reply Source/MDK-Complete.sln:3]: It is Visual Studio doing this afaik. Those are indeed not supposed to be commited (I want to keep a clean ready to test commit for me but I don't know yet how to remove some on the files for the PR candidate.

**Comment by malware-dev** [comment]
```cs
abstract class Block
{
    public List<Block> Children { get; } = new();
    public abstract void Evaluate(IImmutableSet<string> macros, StringBuilder result);
}
```
This is the base class for all blocks. You will have to change the macros list into a _mutable_ list, and the Define block - when evaluated - has to add the symbol to that list (only once).

File: Source/Mdk.CommandLine/IngameScript/Pack/DefaultProcessors/PreprocessorConditionals.cs [hunk 2]
```diff
@@ -69,6 +69,8 @@ public async Task<Document> ProcessAsync(Document document, IPackContext context
                 //since "#define " is 8 in length
```
Comment (malware-dev) [inline Source/Mdk.CommandLine/IngameScript/Pack/DefaultProcessors/PreprocessorConditionals.cs:69]: You don't need to parse the value yourself. Take a look at how the #if is parsed. It's already parsed for you, you only need to take the first token and sanity-check it.

File: Source/Mdk.CommandLine/IngameScript/Pack/DefaultProcessors/PreprocessorConditionals.cs [hunk 3]
```diff
@@ -69,6 +69,8 @@ public async Task<Document> ProcessAsync(Document document, IPackContext context
...
                 string extractedDefineString = tmpStr.Substring(8);
                 allExistingLocalDefineBuilder.Add(extractedDefineString);
```
Comment (malware-dev) [inline Source/Mdk.CommandLine/IngameScript/Pack/DefaultProcessors/PreprocessorConditionals.cs:71]: This shouldn't be necessary anymore at all.

File: Source/Mdk.CommandLine/IngameScript/Pack/DefaultProcessors/PreprocessorConditionals.cs [hunk 4]
```diff
@@ -33,6 +33,18 @@ public async Task<Document> ProcessAsync(Document document, IPackContext context
         Stack<Block> stack = new();
         stack.Push(root);
         var needsMacroCheck = false;
+
+
+        ImmutableHashSet<string>.Builder allExistingLocalDefineBuilder = ImmutableHashSet.CreateBuilder<string>();
```
Comment (malware-dev) [inline Source/Mdk.CommandLine/IngameScript/Pack/DefaultProcessors/PreprocessorConditionals.cs:38]: You just need the allSymbols list. You probably use a HashSet<string> in place of a list for quick duplicate rejection. Either way there's no need for multiple lists here, everything must end up in the same place anyway.

File: Source/Mdk.CommandLine/IngameScript/Pack/DefaultProcessors/PreprocessorConditionals.cs [hunk 5]
```diff
@@ -33,6 +33,18 @@ public async Task<Document> ProcessAsync(Document document, IPackContext context
...
+
+        if(context.PreprocessorSymbols != ImmutableHashSet<string>.Empty)
+        {
+            foreach (var defineContexts in context.PreprocessorSymbols)
+            {
+                allExistingLocalDefineBuilder.Add(defineContexts);
```
Comment (malware-dev) [inline Source/Mdk.CommandLine/IngameScript/Pack/DefaultProcessors/PreprocessorConditionals.cs:44]: As mentioned, add these to the allSymbols list

File: Source/Mdk.CommandLine/IngameScript/Pack/DefaultProcessors/PreprocessorConditionals.cs [hunk 6]
```diff
@@ -95,8 +119,12 @@ public async Task<Document> ProcessAsync(Document document, IPackContext context
             stack.Peek().Children.Add(textBlock);
         }
 
+
+        List<string> allSymbols = allExistingLocalDefineBuilder.ToList(); ;
+
         var result = new StringBuilder();
-        root.Evaluate(context.PreprocessorSymbols ?? ImmutableHashSet<string>.Empty, result);
+
+        root.Evaluate(allSymbols ?? new List<string>(), result);
```
Comment (malware-dev) [inline Source/Mdk.CommandLine/IngameScript/Pack/DefaultProcessors/PreprocessorConditionals.cs:127]: allSymbols will never be null when you've made the above changes

**Comment by gg64du02** [comment]
Ok, you may take a look now. See if it is fit enough.

File: Source/Mdk.CommandLine/IngameScript/Pack/DefaultProcessors/PreprocessorConditionals.cs [hunk 7]
```diff
@@ -64,12 +67,8 @@ public async Task<Document> ProcessAsync(Document document, IPackContext context
...
             if (tokens[0].Kind == Kind.Define)
             {
-                TextLine textLineTL = line;
-                string tmpStr = textLineTL.ToString();
-                //since "#define " is 8 in length
-                string extractedDefineString = tmpStr.Substring(8);
-                allExistingLocalDefineBuilder.Add(extractedDefineString);
-                var defineBlock = new DefineBlock(extractedDefineString);
+                var defineBlock = new DefineBlock(tokens.Skip(1).ToImmutableArray());
+                allSymbols.Add(defineBlock.symbolName);
```
Comment (malware-dev) [inline Source/Mdk.CommandLine/IngameScript/Pack/DefaultProcessors/PreprocessorConditionals.cs:71]: This isn't necessary, is it? The block is adding it.

**Review by malware-dev — APPROVED** [review]

**Pull request merged**